homeassistant/components/esphome/__init__.py

```diff
@@ -777,29 +812,43 @@ def _wrapper(self):
             return None
         return val
 
-    return _wrapper
+    return cast(_PropT, _wrapper)
+
+
+_EnumT = TypeVar("_EnumT", bound=APIIntEnum)
+_ValT = TypeVar("_ValT")
 
 
-class EsphomeEnumMapper(Generic[_T]):
+class EsphomeEnumMapper(Generic[_EnumT, _ValT]):
     """Helper class to convert between hass and esphome enum values."""
 
-    def __init__(self, mapping: dict[_T, str]) -> None:
+    def __init__(self, mapping: dict[_EnumT, _ValT]) -> None:
         """Construct a EsphomeEnumMapper."""
         # Add none mapping
-        mapping = {None: None, **mapping}
-        self._mapping = mapping
-        self._inverse: dict[str, _T] = {v: k for k, v in mapping.items()}
+        augmented_mapping: dict[_EnumT | None, _ValT | None] = mapping  # type: ignore[assignment]
+        augmented_mapping[None] = None
+
+        self._mapping = augmented_mapping
+        self._inverse: dict[_ValT, _EnumT] = {v: k for k, v in mapping.items()}
+
+    @overload
+    def from_esphome(self, value: _EnumT) -> _ValT:
+        ...
+
+    @overload
+    def from_esphome(self, value: _EnumT | None) -> _ValT | None:
```

Comment:
Is this overload needed if it's the same as the one below?

Reply:
It is - with mypy only the _overloads_ are considered for typing information (not the actual declaration). Error without this overload:

```
$ mypy homeassistant/components/esphome/
homeassistant/components/esphome/__init__.py:834: error: Single overload definition, multiple required  [misc]
```